### PR TITLE
Add cases for subcommand rebase of qemu-img

### PR DESCRIPTION
**Pull request by pingl2016**
1. Add cases for unsafe mode/cache mode/show progress
2. Add case for iscsi backend

Signed-off-by: Ping Li pingl@redhat.com

ID: 1361861, 1362614, 1368643

### CODE REVIEW

**Comment by pingl2016** [comment]
Result for qemu_disk_img.rebase.snB.to_base.cache_mode:
(1/6) smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.rebase.snB.to_base.cache_mode.none: PASS (424.44 s)
 (2/6) smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.rebase.snB.to_base.cache_mode.wirteback: PASS (393.87 s)
 (3/6) smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.rebase.snB.to_base.cache_mode.unsafe: PASS (388.86 s)
 (4/6) smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.rebase.snB.to_base.cache_mode.directsync: PASS (396.35 s)
 (5/6) smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.rebase.snB.to_base.cache_mode.writethrough: PASS (388.13 s)

Result for qemu_img.rebase:
(1/5) smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_img.rebase.unsafe: PASS (69.08 s)
 (2/5) smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_img.rebase.show_progress.show_progress_on: PASS (143.11 s)
 (3/5) smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_img.rebase.show_progress.show_progress_off: PASS (150.85 s)
 (4/5) smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_img.rebase.new_to_base.on_iscsi: PASS (189.78 s)

File: qemu/tests/cfg/qemu_disk_img.cfg [hunk 1]
```diff
@@ -168,6 +168,20 @@
                     variants:
                         - to_base:
                             rebase_list = "snB > image1"
+                            variants:
+                                - @default:
```
Comment (suqinhuang) [inline qemu/tests/cfg/qemu_disk_img.cfg]: does default = none ?

File: qemu/tests/cfg/qemu_disk_img.cfg [hunk 2]
```diff
@@ -168,6 +168,20 @@
...
+                                        - none:
+                                            cache_mode = none
+                                        - wirteback:
+                                            cache_mode = writeback
+                                        - unsafe:
+                                            cache_mode = unsafe
+                                        - directsync:
+                                            cache_mode = directsync
+                                        - writethrough:
+                                            cache_mode = writethrough
```
Comment (suqinhuang) [inline qemu/tests/cfg/qemu_disk_img.cfg]: cache_mode used in several sub-test, suggest to create an independent variants

**Comment by pingl2016** [comment]
based on avocado-framework/avocado-vt#698

**Comment by weliao** [comment]
11:23:20 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-41.0-py2.7.egg/avocado_vt/test.py:420
11:23:20 ERROR| Traceback (most recent call last):
11:23:20 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-41.0-py2.7.egg/avocado_vt/test.py", line 196, in runTest
11:23:20 ERROR|     self._runTest()
11:23:20 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-41.0-py2.7.egg/avocado_vt/test.py", line 337, in _runTest
11:23:20 ERROR|     run_func(self, params, env)
11:23:20 ERROR|   File "/usr/lib/python2.7/site-packages/autotest/client/shared/error.py", line 140, in new_fn
11:23:20 ERROR|     return fn(*args, **kwargs)
11:23:20 ERROR|   File "//home/tp-qemu/qemu/tests/qemu_img.py", line 710, in run
11:23:20 ERROR|     eval("%s_test(cmd)" % subcommand)
11:23:20 ERROR|   File "<string>", line 1, in <module>
11:23:20 ERROR|   File "//home/tp-qemu/qemu/tests/qemu_img.py", line 550, in rebase_test
11:23:20 ERROR|     _create(cmd, sn1, sn_fmt, None, eval(backing_file_snapshot1), base_fmt)
11:23:20 ERROR|   File "<string>", line 1, in <module>
11:23:20 ERROR| NameError: name 'base_img' is not defined
11:23:20 ERROR|
11:23:20 ERROR| Traceback (most recent call last):

11:23:20 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-41.0-py2.7.egg/avocado_vt/test.py", line 434, in _run_avocado
    raise test_exception

11:23:20 ERROR| NameError: name 'base_img' is not defined

11:23:20 ERROR| ERROR 02-smp_4.4096m.run_test.Host_RHEL.m7.u3.spice.qcow2.virtio_blk.up.virtio_net.RHEL.6.7.x86_64.my-provider.qemu_img.rebase.unsafe -> NameError: name 'base_img' is not defined

**Comment by pingl2016** [comment]
@weliao Do you update the configuration of iscsi server in the cfg file?

File: qemu/tests/cfg/qemu_img.cfg [hunk 3]
```diff
@@ -90,9 +90,39 @@
...
+                            show_progress = off
+                            check_output = stderr
+                            cache_mode = directsync
+                - new_to_base:
+                    backing_file_snapshot2 = ""
+                    check_output = stderr
+                    command_result_pattern = "Could not open old backing file ''"
+                    skip_boot = yes
+                    variants:
+                        - on_iscsi:
```
Comment (suqinhuang) [inline qemu/tests/cfg/qemu_img.cfg]: there will be more scenerios under new_to_base?
if no, you can update to new_to_base_iscsi:

File: qemu/tests/qemu_img.py [hunk 4]
```diff
@@ -509,40 +516,72 @@ def rebase_test(cmd):
...
+            if "sn1" in lv_name:
+                sn1 = lv
+            if "sn2" in lv_name:
+                sn2 = lv
+        if backing_file_snapshot1 == "":
+            _create(cmd, sn1, sn_fmt, lv_size)
+        else:
+            _create(cmd, sn1, sn_fmt, None, eval(backing_file_snapshot1), base_fmt)
+        if backing_file_snapshot2 == "":
+            _create(cmd, sn2, sn_fmt, lv_size)
```
Comment (suqinhuang) [inline qemu/tests/qemu_img.py]: can we use create_image in the framework directly?

File: qemu/tests/qemu_img.py [hunk 5]
```diff
@@ -483,16 +484,22 @@ def _rebase(cmd, img_name, base_img, backing_fmt, mode="unsafe"):
         :param mode: rebase mode: safe mode, unsafe mode
         """
         cmd += " rebase"
+        show_progress = params.get("show_progress", "")
+        cache_mode = params.get("cache_mode", "")
         if mode == "unsafe":
             cmd += " -u"
+        if show_progress == "on":
```
Comment (suqinhuang) [inline qemu/tests/qemu_img.py]: is it possible to put the cmd to framework ?

File: qemu/tests/cfg/qemu_img.cfg (hunk 3 above)
Comment (pingl2016) [inline qemu/tests/cfg/qemu_img.cfg]: No more scenarios now. Add a separate variants is for expanding case in the future.

File: qemu/tests/qemu_img.py (hunk 5 above)
Comment (pingl2016) [inline qemu/tests/qemu_img.py]: The option is hardly used for qemu-img.

File: qemu/tests/cfg/qemu_img.cfg (hunk 3 above)
Comment (xiagao) [inline qemu/tests/cfg/qemu_img.cfg]: I think it's not clear to use "new_to_base",  “new” is not suitable.

Comment (pingl2016) [inline qemu/tests/cfg/qemu_img.cfg]: here we just create an image without backing file. For the compatibility of cases in qemu_disk_img.cfg, I think "new" is suitable.

**Comment by weliao** [comment]
(01/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_img.rebase.unsafe:  PASS (51.21 s)
 (02/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_img.rebase.show_progress.show_progress_on:  PASS (107.16 s)
 (03/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_img.rebase.show_progress.show_progress_off:  PASS (99.05 s)
 (04/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_img.rebase.new_to_base.on_iscsi:  PASS (374.59 s)
 (05/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snB.to_base:  PASS (351.10 s)
 (06/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snB.to_base.cache_mode.none:  PASS (255.08 s)
 (07/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snB.to_base.cache_mode.wirteback:  PASS (257.81 s)
 (08/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snB.to_base.cache_mode.unsafe:  PASS (250.76 s)
 (09/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snB.to_base.cache_mode.directsync:  PASS (249.69 s)
 (10/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snB.to_base.cache_mode.writethrough:  PASS (254.03 s)
 (11/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snB.to_null:  PASS (289.07 s)
 (12/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snC.to_base:  PASS (305.20 s)
 (13/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snC.to_new:  PASS (346.00 s)
 (14/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snC.to_new.restore_to_snB:  PASS (400.31 s)
 (15/32) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.rebase.snD.to_snB_snA_base:  PASS (576.79 s)

File: qemu/tests/qemu_img.py [hunk 6]
```diff
@@ -509,40 +516,75 @@ def rebase_test(cmd):
...
-        _create(cmd, sn2, sn_fmt, base_img=sn1, base_img_fmt=sn_fmt)
+        skip_boot = params.get("skip_boot", "no")
+        lv_name_list = params.get("lv_name_list", "")
+        vg_name = params.get("vg_name", "")
+        backing_file_snapshot1 = params.get("backing_file_snapshot1", "base_img")
+        backing_file_snapshot2 = params.get("backing_file_snapshot2", "sn1")
+        sn1 = params.get("image_name_snapshot1", "images/sn1")
+        sn2 = params.get("image_name_snapshot2", "images/sn2")
+        if lv_name_list:
+            lv_size = params.get("lv_size", "")
```
Comment (xiagao) [inline qemu/tests/qemu_img.py]: It's better to add steps info here , such as "setup lvmdevice: lv_base,lv_sn1,lv_sn2" ;

**Pull request closed without merging**